controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+                    request.response_tx.send(Ok(response)).unwrap();
+                } else {
+                    // We have no outstanding request.
+                    //
+                    // There are a lot of reasons this might be the case, such
+                    // as a duplicate response from the SP for a previous
+                    // request. It's not obvious what to do here, but for now,
+                    // let's log and drop the message.
+                    warn!(
+                        log,
+                        "received response without an outstanding request";
+                        "message" => ?message,
+                        "peer" => peer,
+                    );
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+async fn send_protocol_error(
+    log: &Logger,
+    socket: &UdpSocket,
+    peer: &SocketAddr,
+    header: &Header,
+    modules: &ModuleId,
+) {
+    error!(log, "Protocol error");
+    let err = MessageError::ProtocolError;
+    let message = Message {
+        header: *header,
+        modules: *modules,
+        body: MessageBody::HostResponse(HostResponse::Error(err)),
+    };
+    let mut buf = [0u8; Message::MAX_SIZE];
+    hubpack::serialize(&mut buf, &message).unwrap();
+    match socket.send_to(&buf, peer).await {
+        Err(e) => {
+            error!(
+                log,
+                "failed to send protocol error";
+                "reason" => ?e,
+                "peer" => peer
+            );
+        }
+        Ok(n_bytes) => assert_eq!(n_bytes, Message::MAX_SIZE),
+    }
+}
+
+async fn send_version_mismatch(
+    log: &Logger,
+    socket: &UdpSocket,
+    peer: &SocketAddr,
+    header: &Header,
+    modules: &ModuleId,
+) -> Result<(), Error> {
+    error!(
+        log,
+        "Mismatched protocol versions";
+        "expected" => message::version::V1,
+        "actual" => header.version,
+    );
+    let err = MessageError::VersionMismatch {
+        expected: message::version::V1,
+        actual: header.version,
+    };
+    let message = Message {
+        header: *header,
+        modules: *modules,
+        body: MessageBody::HostResponse(HostResponse::Error(err)),
+    };
+    let mut buf = [0u8; Message::MAX_SIZE];
+    hubpack::serialize(&mut buf, &message).unwrap();
+    let n_bytes = socket.send_to(&buf, peer).await?;
+    assert_eq!(n_bytes, Message::MAX_SIZE);
+    Ok(())
+}
+
+// Send a request from the host to the SP, returning the number of bytes sent.
+async fn send_outgoing_request(
+    socket: &UdpSocket,
+    peer: &SocketAddrV6,
+    request: &mut OutstandingHostRequest,
+    buf: &mut [u8],
+) -> Result<usize, Error> {
+    let mut msg_size = match hubpack::serialize(buf, &request.request.message) {
+        Ok(n) => n,
+        Err(e) => return Err(Error::SerDes(e)),
+    };
+    if let Some(data) = &request.request.data {
+        buf[msg_size..data.len()].copy_from_slice(data);
```

Comment:
There's a bug here. The end should be `msg_size + data.len()`.

Reply:
Bit of clever syntax I think I picked up from @mkeeter for cases like this:

```rust
buf[msg_size..][..data.len()].copy_from_slice(data);
```

to avoid having to remember to do the addition.

Reply:
Do we need to check and return an error if `data` doesn't run past the end of `buf`? Currently this panics, which is maybe fine if we know buf is always big enough.

Reply:
That's a good question. I've not really written the part of the `Controller` that writes data, which I think is the only one accepting data on the host side at this point. But I was planning to have a function `Controller::write()`, which accepts `data` and returns an error if the data is too big to ever be serialized. I'm not sure how that'll play out, but that would mean that we should never overrun the buffer.